In [1]:
%load_ext sql

In [2]:
from sqlalchemy import create_engine
import psycopg2
import pandas as pd
import numpy as np
from scipy import stats
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_validate
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn import svm
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import Perceptron

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
%sql postgresql://zgbodaynfjorzt:d2b11c0c6039df67f39d22f680fb92a843a1b652e3d5b692f6670c0a2abc2804@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm

In [5]:
# Format
engine = create_engine('postgresql://zgbodaynfjorzt:d2b11c0c6039df67f39d22f680fb92a843a1b652e3d5b692f6670c0a2abc2804@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm')

In [6]:
engine

Engine(postgresql://zgbodaynfjorzt:***@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm)

In [7]:
tables_list = engine.table_names()
tables_list

/tmp/ipykernel_278991/431584093.py:1: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)
  tables_list = engine.table_names()


['forms', 'order_items', 'orders']

In [8]:
tables_list[0]

'forms'

In [9]:
%%sql

SELECT *
FROM forms
limit 1

 * postgresql://zgbodaynfjorzt:***@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm
1 rows affected.


[('c7d453c9c01223a42968d9fe136047', 'd2620a3e6f00a6dc6e88', datetime.datetime(2022, 1, 3, 15, 42, 22), 'iOS', 'Safari', 'São Paulo, Sao Paulo', 'BR', 'SÃO PAULO', ' SAO PAULO', 4, None, 'Tipo 2', 'Raíz exposta', 'Pontas duplas', 'Nenhum', 'Há 2 semanas', 'Celíaca', 'Frequente', 'Nunca', 'Característica 1', '25 a 34', 'Efeito 7', 'Rosas')]

In [10]:
%%sql

SELECT *
FROM order_items
limit 1

 * postgresql://zgbodaynfjorzt:***@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm
1 rows affected.


[(423803, 'dd7dd09eb757eab', '62296554a44f818ced0d623434e26b', '509d', 2, Decimal('74.95'))]

In [11]:
%%sql

SELECT *
FROM orders
limit 1

 * postgresql://zgbodaynfjorzt:***@ec2-3-222-74-92.compute-1.amazonaws.com/dfo2vqe9hs2knm
1 rows affected.


[('c677716e0f2901f', '001b7aed7554265e4e5f', datetime.datetime(2022, 3, 22, 21, 10, 49), 'Entregue', 'CARTÃO', Decimal('198.63'), Decimal('0.00'), True, 'Osasco', 'SP', 'BR', 'google')]

In [12]:
pd.read_sql_query('''SELECT * FROM forms LIMIT 1;''', engine) 

id_form            id_cliente                data  \
0  c7d453c9c01223a42968d9fe136047  d2620a3e6f00a6dc6e88 2022-01-03 15:42:22   

  sistema_operacional navegador           localizacao locale     cidade  \
0                 iOS    Safari  São Paulo, Sao Paulo     BR  SÃO PAULO   

       estado  qtde_lavagem_semana utm_source tipo_cabelo   comprimento  \
0   SAO PAULO                    4       None      Tipo 2  Raíz exposta   

       tipo_fios procedimentos tempo_procedimento    dieta atividade_fisica  \
0  Pontas duplas        Nenhum       Há 2 semanas  Celíaca        Frequente   

  frequencia_estresse    caracteristica faixa_etaria efeitos_desejados  \
0               Nunca  Característica 1      25 a 34          Efeito 7   

  fragancia  
0     Rosas

### A empresa está gastando muito dinheiro com disparos via Whatsapp para pessoas que preenchem o formulário (leads), desta forma seria interessante qualificar quem são os leads que possuem maior probabilidade de efetuar uma compra e assim tornar os disparos mais efetivos. 

### Achar a correlação entre compradores e leads. Para isso, acesse o banco de dados presente em sua empresa para fazer a coleta dos dados, e use os recursos ao seu alcance para solucionar o problema e apresentar as suas conclusões.

### 1. Leitura dos dados e normalização 

In [13]:
df_forms = pd.read_sql_query('''SELECT * FROM forms''', engine)

In [14]:
df_forms.head()

id_form            id_cliente                data  \
0  c7d453c9c01223a42968d9fe136047  d2620a3e6f00a6dc6e88 2022-01-03 15:42:22   
1  72830da153824af486be15cd48b3f3  9a48802d8290d19471f3 2022-01-04 08:39:53   
2  d101fc0075dac29db1168cfc61ad44  c3344fad450ee424700d 2022-01-04 19:34:05   
3  0957fb9d7be36b4d079b6582bb6785  8105e34124e90300d8b5 2022-01-05 18:29:41   
4  ba73bdcc6c063ab08b533eae5505eb  636382f3f545eec3fe97 2022-01-07 11:44:31   

  sistema_operacional navegador                localizacao locale  \
0                 iOS    Safari       São Paulo, Sao Paulo     BR   
1             Windows    Chrome  Ribeirão Preto, Sao Paulo     BR   
2             Android    Chrome       São Paulo, Sao Paulo     BR   
3             Android    Chrome        Manchester, England     BR   
4             Android    Chrome    Tijucas, Santa Catarina     BR   

           cidade           estado  qtde_lavagem_semana utm_source  \
0       SÃO PAULO        SAO PAULO                    4       None   
1  RIBEIRÃO PRETO        SAO PAULO                    7       None   
2       SÃO PAULO        SAO PAULO                    2       None   
3      MANCHESTER          ENGLAND                    5       None   
4         TIJUCAS   SANTA CATARINA                    4   facebook   

  tipo_cabelo   comprimento      tipo_fios procedimentos tempo_procedimento  \
0      Tipo 2  Raíz exposta  Pontas duplas        Nenhum       Há 2 semanas   
1      Tipo 4   Muito curto      Perfeitos   Todos acima   Há mais de 1 mês   
2      Tipo 1   Extra longo       Rebeldes   Todos acima   Há mais de 1 mês   
3      Tipo 1   Extra longo       Rebeldes   Todos acima   Há mais de 1 mês   
4      Tipo 3   Muito curto       Rebeldes       Tintura        Há 1 semana   

               dieta atividade_fisica frequencia_estresse     caracteristica  \
0            Celíaca        Frequente               Nunca   Característica 1   
1  Nenhuma restrição            Nunca               Nunca   Característica 6   
2  Nenhuma restrição        Raramente        Regularmente   Característica 7   
3            Celíaca        Raramente        Regularmente   Característica 7   
4            Celíaca           Sempre               Nunca  Característica 10   

  faixa_etaria efeitos_desejados  fragancia  
0      25 a 34          Efeito 7      Rosas  
1      55 a 64          Efeito 1  Eucalipto  
2       2 a 10          Efeito 7      Rosas  
3      25 a 34          Efeito 7      Rosas  
4      18 a 24          Efeito 7  Eucalipto

In [15]:
len(df_forms)

427950

In [16]:
df_order_items = pd.read_sql_query('''SELECT * FROM order_items''', engine)

In [17]:
df_order_items.head()

id        id_pedido                         id_form id_produto  qtde  \
0  423803  dd7dd09eb757eab  62296554a44f818ced0d623434e26b       509d     2   
1  423804  dd7dd09eb757eab  62296554a44f818ced0d623434e26b       e820     1   
2  423805  dd7dd09eb757eab  62296554a44f818ced0d623434e26b       d346     2   
3  423806  dd7dd09eb757eab  62296554a44f818ced0d623434e26b       d2e0     1   
4  423807  46a72a1e08c387e  25e1f26f2ec67b0cbefde74337185a       509d     2   

   valor  
0  74.95  
1  74.95  
2  59.90  
3  64.90  
4  74.95

In [18]:
df_orders = pd.read_sql_query('''SELECT * FROM orders''', engine)

In [19]:
df_orders.head()

id_pedido            id_cliente         data_pedido status_pagamento  \
0  c677716e0f2901f  001b7aed7554265e4e5f 2022-03-22 21:10:49         Entregue   
1  df063dd3f6b1641  d107878c463e7c969ce6 2022-03-22 21:47:08         Entregue   
2  dd69f88884e45a8  44845cf79366dbcecf00 2022-03-08 22:50:22         Entregue   
3  2c556256ec3a77e  2aa1a61cfa7be6c51167 2022-01-21 14:53:44         Entregue   
4  0262be22fc1c6f0  10c138383539f471831f 2022-03-08 22:50:24         Entregue   

  tipo_pagamento  valor_total  valor_desconto  frete_gratis      cidade  \
0         CARTÃO       198.63             0.0          True      Osasco   
1         CARTÃO       198.63             0.0          True   São Paulo   
2         CARTÃO       260.96             0.0          True       Bauru   
3         CARTÃO       225.24             0.0         False  Ananindeua   
4         CARTÃO       199.31             0.0          True   Saquarema   

  estado locale  utm_source  
0     SP     BR      google  
1     SP     BR  influencer  
2     SP     BR    facebook  
3     PA     BR    facebook  
4     RJ     BR    facebook

In [20]:
len(df_orders)

55078

####  Pegar a relação de todos os id_pedido em df_orders, ver qual é o id_form em df_order_items e ver quais desses aparecem em df_forms

In [21]:
# Relação de todos os pedidos feitos:
id_pedido = df_orders['id_pedido']

In [22]:
id_pedido.head()

0    c677716e0f2901f
1    df063dd3f6b1641
2    dd69f88884e45a8
3    2c556256ec3a77e
4    0262be22fc1c6f0
Name: id_pedido, dtype: object

In [23]:
# Dado estes id_pedido, obter o respectivo id_form em df_order_items:
df_temp = df_order_items[df_order_items['id_pedido'].isin(id_pedido)]
id_form = df_temp['id_form']

In [24]:
id_form.head()

0    62296554a44f818ced0d623434e26b
1    62296554a44f818ced0d623434e26b
2    62296554a44f818ced0d623434e26b
3    62296554a44f818ced0d623434e26b
4    25e1f26f2ec67b0cbefde74337185a
Name: id_form, dtype: object

In [25]:
id_form_list = id_form.tolist()

In [26]:
# Dado estes id_form, foi criado uma coluna 'class' com o valor True para quem é comprador e False para quem não é.
df_forms['class'] = df_forms['id_form'].isin(id_form_list)

In [27]:
# Convertendo o booleando para inteiro:
df_forms['class'] = df_forms['class'].astype(int)

In [28]:
df_forms.head()

id_form            id_cliente                data  \
0  c7d453c9c01223a42968d9fe136047  d2620a3e6f00a6dc6e88 2022-01-03 15:42:22   
1  72830da153824af486be15cd48b3f3  9a48802d8290d19471f3 2022-01-04 08:39:53   
2  d101fc0075dac29db1168cfc61ad44  c3344fad450ee424700d 2022-01-04 19:34:05   
3  0957fb9d7be36b4d079b6582bb6785  8105e34124e90300d8b5 2022-01-05 18:29:41   
4  ba73bdcc6c063ab08b533eae5505eb  636382f3f545eec3fe97 2022-01-07 11:44:31   

  sistema_operacional navegador                localizacao locale  \
0                 iOS    Safari       São Paulo, Sao Paulo     BR   
1             Windows    Chrome  Ribeirão Preto, Sao Paulo     BR   
2             Android    Chrome       São Paulo, Sao Paulo     BR   
3             Android    Chrome        Manchester, England     BR   
4             Android    Chrome    Tijucas, Santa Catarina     BR   

           cidade           estado  qtde_lavagem_semana utm_source  \
0       SÃO PAULO        SAO PAULO                    4       None   
1  RIBEIRÃO PRETO        SAO PAULO                    7       None   
2       SÃO PAULO        SAO PAULO                    2       None   
3      MANCHESTER          ENGLAND                    5       None   
4         TIJUCAS   SANTA CATARINA                    4   facebook   

  tipo_cabelo   comprimento      tipo_fios procedimentos tempo_procedimento  \
0      Tipo 2  Raíz exposta  Pontas duplas        Nenhum       Há 2 semanas   
1      Tipo 4   Muito curto      Perfeitos   Todos acima   Há mais de 1 mês   
2      Tipo 1   Extra longo       Rebeldes   Todos acima   Há mais de 1 mês   
3      Tipo 1   Extra longo       Rebeldes   Todos acima   Há mais de 1 mês   
4      Tipo 3   Muito curto       Rebeldes       Tintura        Há 1 semana   

               dieta atividade_fisica frequencia_estresse     caracteristica  \
0            Celíaca        Frequente               Nunca   Característica 1   
1  Nenhuma restrição            Nunca               Nunca   Característica 6   
2  Nenhuma restrição        Raramente        Regularmente   Característica 7   
3            Celíaca        Raramente        Regularmente   Característica 7   
4            Celíaca           Sempre               Nunca  Característica 10   

  faixa_etaria efeitos_desejados  fragancia  class  
0      25 a 34          Efeito 7      Rosas      1  
1      55 a 64          Efeito 1  Eucalipto      0  
2       2 a 10          Efeito 7      Rosas      0  
3      25 a 34          Efeito 7      Rosas      0  
4      18 a 24          Efeito 7  Eucalipto      0

Conversão de variável categórica em um inteiro. Categorizar do df_forms as colunas:

    - 'qtde_lavagem_semana'
    - 'tipo_cabelo'
    - 'comprimento'
    - 'procedimentos'
    - 'tempo_procedimento'
    - 'dieta'
    - 'atividade física'
    - 'frequencia_estresse'
    - 'caracteristica'
    - 'faixa_etaria'
    - 'efeitos_desejados'
    - 'fragancia'

In [29]:
df_features = df_forms[['qtde_lavagem_semana', 'tipo_cabelo', 'comprimento', 'procedimentos', 
                        'tempo_procedimento', 'dieta', 'atividade_fisica', 'frequencia_estresse', 
                        'caracteristica', 'faixa_etaria', 'efeitos_desejados', 'fragancia']]

In [30]:
df_features.head()

qtde_lavagem_semana tipo_cabelo   comprimento procedimentos  \
0                    4      Tipo 2  Raíz exposta        Nenhum   
1                    7      Tipo 4   Muito curto   Todos acima   
2                    2      Tipo 1   Extra longo   Todos acima   
3                    5      Tipo 1   Extra longo   Todos acima   
4                    4      Tipo 3   Muito curto       Tintura   

  tempo_procedimento              dieta atividade_fisica frequencia_estresse  \
0       Há 2 semanas            Celíaca        Frequente               Nunca   
1   Há mais de 1 mês  Nenhuma restrição            Nunca               Nunca   
2   Há mais de 1 mês  Nenhuma restrição        Raramente        Regularmente   
3   Há mais de 1 mês            Celíaca        Raramente        Regularmente   
4        Há 1 semana            Celíaca           Sempre               Nunca   

      caracteristica faixa_etaria efeitos_desejados  fragancia  
0   Característica 1      25 a 34          Efeito 7      Rosas  
1   Característica 6      55 a 64          Efeito 1  Eucalipto  
2   Característica 7       2 a 10          Efeito 7      Rosas  
3   Característica 7      25 a 34          Efeito 7      Rosas  
4  Característica 10      18 a 24          Efeito 7  Eucalipto

In [31]:
df_features_transformed = df_features.apply(LabelEncoder().fit_transform)

In [32]:
# Standardizing the features
# df_features_transformed = StandardScaler().fit_transform(df_features_transformed)
df_features_transformed = stats.zscore(df_features_transformed)

In [33]:
df_features_transformed

qtde_lavagem_semana  tipo_cabelo  comprimento  procedimentos  \
0                  0.470162     0.013822     1.064319       0.032408   
1                  2.158373     1.857434    -0.170651       1.348807   
2                 -0.655311    -0.907984    -2.640589       1.348807   
3                  1.032899    -0.907984    -2.640589       1.348807   
4                  0.470162     0.935628    -0.170651       0.690607   
...                     ...          ...          ...            ...   
427945             0.470162    -0.907984    -0.170651       0.032408   
427946            -1.218048    -0.907984     1.064319       0.690607   
427947             2.158373    -0.907984    -0.170651       0.032408   
427948             0.470162    -0.907984     1.064319       1.348807   
427949             1.595636     0.013822    -0.170651      -0.214417   

        tempo_procedimento     dieta  atividade_fisica  frequencia_estresse  \
0                 0.101530 -0.575831         -1.464914            -1.184785   
1                 0.986832  1.877082         -0.078154            -1.184785   
2                 0.986832  1.877082          0.615227             0.202260   
3                 0.986832 -0.575831          0.615227             0.202260   
4                -0.783772 -0.575831          1.308607            -1.184785   
...                    ...       ...               ...                  ...   
427945           -0.783772  1.877082         -1.464914             0.895782   
427946           -1.669073 -0.575831          0.615227             0.202260   
427947            0.986832  1.877082          1.308607            -1.184785   
427948            0.986832  1.877082          1.308607            -1.184785   
427949           -0.783772 -0.575831          0.615227            -1.184785   

        caracteristica  faixa_etaria  efeitos_desejados  fragancia  
0            -1.757383      0.287762           0.579041   0.921563  
1             1.137187      2.238779          -2.886497  -1.205695  
2             1.404021     -0.362578           0.579041   0.921563  
3             1.404021      0.287762           0.579041   0.921563  
4            -0.939479     -1.012917           0.579041  -1.205695  
...                ...           ...                ...        ...  
427945        1.404021     -1.012917           0.579041   0.921563  
427946        1.755932      2.889118           0.624332  -0.142066  
427947        1.137187      2.238779          -2.886497  -1.205695  
427948        1.405954     -1.012917           0.636032  -1.205695  
427949       -0.674578      1.588440           1.173111   0.921563  

[427950 rows x 12 columns]

Foi identificado inconsistências, mas não foi feito o tratamento. 

Por exemplo, está BR para a localidade de um cliente na cidade de Manchester. E nesta mesma linha está England para Estado.

In [34]:
# Salvando em csv

df_forms.to_csv('./forms.zip', index=False)
df_order_items.to_csv('./order_items.zip', index=False)
df_features.to_csv('./features.zip', index=False)
df_features_transformed.to_csv('./features_transformed.zip', index=False)

### 2. Separação de bases para treinar e efetuar o crossvalidation

In [80]:
# Separação dos dados: features e classe
X = df_features_transformed.copy()
y = df_forms['class']

In [81]:
y.value_counts()

0    373809
1     54141
Name: class, dtype: int64

In [82]:
# Obtendo as 5 melhores features (mais pertinentes), dada a classe
from sklearn.decomposition import PCA

In [83]:
pca = PCA(n_components=5)

In [84]:
pca.fit_transform(X)
X_pca = pca.transform(X)
# y = pca.transform(y.reshape(-1, 1))

### 3. Respaldo teórico e prático sobre as escolhas dos algoritmos (podendo ser mais de um)

Foram escolhidos 3 algoritmos de classificação: 

    - Árvore de decisão. A principal vantagem do classificador de árvore de decisão é sua capacidade de usar diferentes subconjuntos de features e regras de decisão em diferentes estágios de classificação, o que permite alta inteligibilidade.
    Fonte: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
    
    - Perceptron. Esta é a rede neural mais simples, com apenas um neurônio, que gera um hyperplano para fazer a classificação de duas classes. É um tipo de classificador linear que faz suas previsões com base em uma função de previsão linear combinando um conjunto de pesos com o vetor de features.
    Fonte: https://en.wikipedia.org/wiki/Perceptron

    - Support Vector Machines, que utiliza um hyperplano e vetores de suporte para separar as classes. O SVM busca o hyperplano mais otimizado, mediante os dados de treino, para esta separação. Este algoritmo é altamente efetivo em espaços de alta dimensionalidade.
    Fonte: https://www.researchgate.net/post/Why-SVM-classifier-is-the-most-powerful-classification-algorithm-specifically-for-binary-classification-task

In [85]:
num_folds = 3

In [86]:
# Decision Trees algorithm

In [87]:
clf1 = DecisionTreeClassifier(max_depth=10,criterion='gini').fit(X_pca, y)

In [88]:
scores1 = cross_val_score(clf1, X_pca, y, cv=num_folds)

In [89]:
print("Decision Trees Accuracy: %0.2f (+/- %0.2f)" % (scores1.mean(), scores1.std() * 2))

Decision Trees Accuracy: 0.87 (+/- 0.00)


In [90]:
# Perceptron Algorithm

In [91]:
clf2 = Perceptron(early_stopping=True, max_iter=1000, shuffle=False).fit(X_pca, y)

In [92]:
scores2 = cross_val_score(clf2, X_pca, y, cv=num_folds)

In [93]:
print("Perceptron Accuracy: %0.2f (+/- %0.2f)" % (scores2.mean(), scores2.std() * 2))

Perceptron Accuracy: 0.62 (+/- 0.70)


In [99]:
# Support Vector Machines algorithm

In [95]:
clf3 = svm.SVC(kernel='linear', C=0.5)

In [96]:
clf3.fit(X_pca, y)

SVC(C=0.5, kernel='linear')

In [97]:
scores3 = cross_val_score(clf3, X_pca, y, cv=num_folds)

In [98]:
print("SVM Accuracy: %0.2f (+/- %0.2f)" % (scores3.mean(), scores3.std() * 2))

SVM Accuracy: 0.87 (+/- 0.00)


### 4. Como e porquê foram escolhidas as features analisadas

In [99]:
n_components = pca.components_.shape[0]

In [100]:
# get the index of the most important feature on EACH component i.e. largest absolute value
most_important = [np.abs(pca.components_[i]).argmax() for i in range(n_components)]

In [101]:
feature_names = ['qtde_lavagem_semana', 'tipo_cabelo', 'comprimento', 'procedimentos', 
                 'tempo_procedimento', 'dieta', 'atividade_fisica', 'frequencia_estresse', 
                 'caracteristica', 'faixa_etaria', 'efeitos_desejados', 'fragancia']

In [102]:
# get the names
most_important_names = [feature_names[most_important[i]] for i in range(n_components)]

In [103]:
dic = {'PC{}'.format(i+1): most_important_names[i] for i in range(n_components)}

In [104]:
# build the dataframe
df = pd.DataFrame(sorted(dic.items()))

In [105]:
df

0                    1
0  PC1   tempo_procedimento
1  PC2                dieta
2  PC3    efeitos_desejados
3  PC4            fragancia
4  PC5  qtde_lavagem_semana

#### De acordo com a técnica Principal Component Analysis (PCA), as 5 features mais pertinentes para análises são estas elencadas acima. O PCA é uma técnica estatística que converte dados de alta dimensionalidae para baixa dimensionalidade ao selecionar as features mais importantes que captura o máximo de informação sobre o conjunto de dadas. As features são selecionadas baseadas na variância que ela causa na saída do modelo. A feature que gera a maior variância é o primeiro componente principal, da mesma forma para a segunda maior variância sendo o segundo componente principal e assim por diante. 

#### Observa-se que as features mais pertinentes são relacionados com o perfil dos compradores, seja relacionada à rotina do cabelo, dieta, bem como o tipo do cabelo.

### 5. Salvar o arquivo de treinamento para uma API consumir e gerar o score

In [106]:
from joblib import dump

In [107]:
dump(clf3, 'model.joblib') 

['model.joblib']

### 6. Resultado final com métricas de acertividade do algoritmo

Os algoritmos Decision Trees e SVM obtiveram os melhores resultados, empatando com 87% de acurácia. Dado que o SVM é um algoritmo custoso, recomenda-se utilizar o Decision Trees. 

Para um estudo mais aprofundado, recomenda-se aumentar o número de folds na validação cruzada e também otimizar os hyperparâmetros para verificar se o modelo consegue uma melhor generalização, sem overfitar. 